opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1121,6 +1167,88 @@ namespace detail {
         }
     }
 
+    template <class Grid, class Implementation>
+    void BlackoilModelBase<Grid, Implementation>::solveWellEq(const std::vector<ADB>& mob_perfcells,
+                                                              const std::vector<ADB>& b_perfcells,
+                                                              SolutionState& state,
+                                                              WellState& well_state)
+    {
+        V aliveWells;
+        const int np = wells().number_of_phases;
+        std::vector<ADB> cq_s(np, ADB::null());
+
+        int          it  = 0;
+        std::vector<V> vars0;
+        //bhp and Q for the wells
+        vars0.reserve(2);
+        variableWellStateInitials(well_state,vars0);
+        std::vector<ADB> vars = ADB::variables(vars0);
+        std::vector<int> indices = variableWellsStateIndices();
+        SolutionState state0 = state;
+        asImpl().makeConstantState(state0);
+        SolutionState wellSolutionState = state0;
+        variableStateExtractWellsVars(indices,vars,wellSolutionState);
+        std::vector<ADB> mob_perfcells_const(np, ADB::null());
+        std::vector<ADB> b_perfcells_const(np, ADB::null());
+        for (int phase = 0; phase < np; ++phase) {
+            mob_perfcells_const[phase] = ADB::constant(mob_perfcells[phase].value());
+            b_perfcells_const[phase] = ADB::constant(b_perfcells[phase].value());
+        }
+        asImpl().addWellEq(wellSolutionState, well_state, mob_perfcells_const, b_perfcells_const, aliveWells,cq_s);
+        addWellControlEq(wellSolutionState, well_state, aliveWells);
+        bool converged = getWellConvergence(it);
+        while ( (!converged && (it < 15))) {
+
+            std::vector<ADB> eqs;
+            eqs.reserve(2);
+            eqs.push_back(residual_.well_flux_eq);
+            eqs.push_back(residual_.well_eq);
+            ADB total_residual = vertcatCollapseJacs(eqs);
+            const std::vector<M>& Jn = total_residual.derivative();
+            const Eigen::SparseLU< M > solver(Jn[0]);
+            const Eigen::VectorXd& dx = solver.solve(total_residual.value().matrix());
+            const int numeq = well_state.numWells()*(well_state.numPhases()+1);
+            V dx_V = V(numeq);
+            std::copy_n(dx.data(),numeq, dx_V.data());
+            updateWellState(dx_V,well_state);
+            updateWellControls(well_state);
+            //bhp and Q for the wells
```

Comment:
The 8 lines here (until the `addWellControlEq()` call) are repeated from above the while. It would be better to use a `do { } while()` loop instead, that way `vars0, vars, wellSolutionState` could all be local to the loop.